File: controllers/gatewayclass_controller.go
/*
Copyright 2023 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package controllers

import (
	"context"

	"github.com/kong/blixt/pkg/vars"
	"k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime"
	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/client"
	"sigs.k8s.io/controller-runtime/pkg/log"
	"sigs.k8s.io/controller-runtime/pkg/predicate"
	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
)

//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses,verbs=get;list;watch;create;update;patch;delete
//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses/status,verbs=get;update;patch
//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses/finalizers,verbs=update

// GatewayClassReconciler reconciles a GatewayClass object
type GatewayClassReconciler struct {
	client.Client
	Scheme *runtime.Scheme
}

// SetupWithManager loads the controller into the provided controller manager.
func (r *GatewayClassReconciler) SetupWithManager(mgr ctrl.Manager) error {
	return ctrl.NewControllerManagedBy(mgr).
		For(&gatewayv1beta1.GatewayClass{}).
		WithEventFilter(predicate.NewPredicateFuncs(func(obj client.Object) bool {
			gwc, ok := obj.(*gatewayv1beta1.GatewayClass)
			if !ok {
				return false
			}
			return gwc.Spec.ControllerName == vars.GatewayClassControllerName // filter out unmanaged GWCs
		})).
		Complete(r)
}

// Reconcile provisions (and de-provisions) resources relevant to this controller.
func (r *GatewayClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
	log := log.FromContext(ctx)

	gwc := new(gatewayv1beta1.GatewayClass)
	if err := r.Client.Get(ctx, req.NamespacedName, gwc); err != nil {
		if errors.IsNotFound(err) {
			log.Info("object enqueued no longer exists, skipping")
			return ctrl.Result{}, nil
		}
		return ctrl.Result{}, err
	}

	if gwc.Spec.ControllerName != vars.GatewayClassControllerName {
		return ctrl.Result{}, nil
	}

	if !r.isAccepted(gwc) {
		log.Info("marking GatwayClass as accepted", "name", gwc.Name)
		return ctrl.Result{}, r.accept(ctx, gwc)
	}

	return ctrl.Result{}, nil

}

func (r *GatewayClassReconciler) isAccepted(gwc *gatewayv1beta1.GatewayClass) bool {
	for _, cond := range gwc.Status.Conditions {
		if cond.Type == string(gatewayv1beta1.GatewayClassConditionStatusAccepted) {
			if cond.Status == metav1.ConditionTrue {
				return true
			}
		}
	}

	return false
}

func (r *GatewayClassReconciler) accept(ctx context.Context, gwc *gatewayv1beta1.GatewayClass) error {
	previousGWC := gwc.DeepCopy()
	acceptedCond := metav1.Condition{
		Type:               string(gatewayv1beta1.GatewayClassConditionStatusAccepted),
		Status:             metav1.ConditionTrue,
		ObservedGeneration: gwc.Generation,
		LastTransitionTime: metav1.Now(),
		Reason:             string(gatewayv1beta1.GatewayClassReasonAccepted),
		Message:            "the gatewayclass has been accepted by the operator",
	}
	setCondition(acceptedCond, gwc)
	return r.Status().Patch(ctx, gwc, client.MergeFrom(previousGWC))
}

func setCondition(condition metav1.Condition, gwc *gatewayv1beta1.GatewayClass) {
	newConds := make([]metav1.Condition, 0, len(gwc.Status.Conditions))

	for i := 0; i < len(gwc.Status.Conditions); i++ {
		if gwc.Status.Conditions[i].Type != condition.Type {
			newConds = append(newConds, gwc.Status.Conditions[i])
		}
	}

	newConds = append(newConds, condition)
	gwc.Status.Conditions = newConds
}

File: controlplane/src/gatewayclass_controller.rs
/*
Copyright 2024 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use futures::StreamExt;
use std::{
    ops::Sub,
    sync::Arc,
    time::{Duration, Instant},
};

use crate::*;
use gateway_api::apis::standard::gatewayclasses::GatewayClass;
use kube::{
    api::{Api, ListParams},
    runtime::{controller::Action, watcher::Config, Controller},
};

use gatewayclass_utils::*;
use tracing::*;

pub async fn reconcile(gateway_class: Arc<GatewayClass>, ctx: Arc<Context>) -> Result<Action> {
    let start = Instant::now();
    let client = ctx.client.clone();
    let name = gateway_class
        .metadata
        .name
        .clone()
        .ok_or(Error::InvalidConfigError("invalid name".to_string()))?;

    let gatewayclass_api = Api::<GatewayClass>::all(client);
    let mut gwc = GatewayClass {
        metadata: gateway_class.metadata.clone(),
        spec: gateway_class.spec.clone(),
        status: gateway_class.status.clone(),
        // NOTE: Am I missing anything else here?
    };

    if gateway_class.spec.controller_name != GATEWAY_CLASS_CONTROLLER_NAME {
        // Skip reconciling because we don't manage this resource
        // NOTE: May want to requeue in case this resource becomes relevant again in the
        // future (e.g. the controllerName is changed to match ours in the event of typo,
        // etc.)
        return Ok(Action::requeue(Duration::from_secs(3600 / 2)));
    }

    if !is_accepted(&gateway_class) {
        info!("marking gateway class {:?} as accepted", name);
        accept(&mut gwc);
        patch_status(&gatewayclass_api, name, &gwc.status.unwrap_or_default()).await?;
    }

    let duration = Instant::now().sub(start);
    info!("finished reconciling in {:?} ms", duration.as_millis());
    Ok(Action::await_change())
}

pub async fn controller(ctx: Context) -> Result<()> {
    let gwc_api = Api::<GatewayClass>::all(ctx.client.clone());
    gwc_api
        .list(&ListParams::default().limit(1))
        .await
        .map_err(Error::CRDNotFoundError)?;

    Controller::new(gwc_api, Config::default().any_semantic())
        .shutdown_on_signal()
        .run(reconcile, error_policy, Arc::new(ctx))
        .filter_map(|x| async move { std::result::Result::ok(x) })
        .for_each(|_| futures::future::ready(()))
        .await;

    Ok(())
}

fn error_policy(_: Arc<GatewayClass>, error: &Error, _: Arc<Context>) -> Action {
    warn!("reconcile failed: {:?}", error);
    Action::requeue(Duration::from_secs(5))
}
